File: android/app/src/main/java/com/naivesound/scherzo/PianoKeyboard.java
package com.naivesound.scherzo;

import android.content.Context;
import android.graphics.Point;
import android.graphics.Rect;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.SparseArray;
import android.util.SparseIntArray;
import android.view.MotionEvent;
import android.view.VelocityTracker;
import android.view.View;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import static trikita.anvil.DSL.*;

import trikita.anvil.Anvil;
import trikita.anvil.RenderableView;

public class PianoKeyboard extends RenderableView {

    private static final String TAG = "PianoKeyboard";

    static boolean[] BLACK_KEYS = new boolean[]{
            false, true, false, true, false, false, true, false, true, false, true, false
    };

    private int mWidth;
    private int mHeight;

    private final Map<View, Set<Integer>> mTouchedViews = new HashMap<>();
    private final SparseIntArray mPrevX = new SparseIntArray();
    private final SparseIntArray mPrevY = new SparseIntArray();

    private int mKeyCount = 12;
    private int mFirstKey = 48;

    public PianoKeyboard(Context c) {
        super(c);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        mWidth = MeasureSpec.getSize(widthMeasureSpec);
        mHeight = MeasureSpec.getSize(heightMeasureSpec);
        this.setMeasuredDimension(mWidth, mHeight);
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        Anvil.render();
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        return true;
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        int action = ev.getActionMasked();
        switch (action) {
            case MotionEvent.ACTION_POINTER_DOWN:
            case MotionEvent.ACTION_DOWN:
                int pointerId = ev.getPointerId(ev.getActionIndex());
                mPrevX.put(pointerId, (int) ev.getX(ev.getActionIndex()));
                mPrevY.put(pointerId, (int) ev.getY(ev.getActionIndex()));
                dispatchTouchDown(findViewByPointer(ev, ev.getActionIndex()), ev);
                break;
            case MotionEvent.ACTION_OUTSIDE:
            case MotionEvent.ACTION_CANCEL:
            case MotionEvent.ACTION_POINTER_UP:
            case MotionEvent.ACTION_UP:
                for (Iterator<View> it = mTouchedViews.keySet().iterator(); it.hasNext(); ) {
                    View v = it.next();
                    if (dispatchTouchUp(v, ev)) {
                        it.remove();
                    }
                }
                break;
            case MotionEvent.ACTION_MOVE:
                for (int i = 0; i < ev.getPointerCount(); i++) {
                    int id = ev.getPointerId(i);
                    int x = (int) ev.getX(i);
                    int y = (int) ev.getY(i);
                    int minDistance = (int) (24 * (getResources().getDisplayMetrics().densityDpi / DisplayMetrics.DENSITY_DEFAULT));
                    if (Math.abs(mPrevX.get(id, x) - x) > minDistance || Math.abs(mPrevY.get(id, y) - y) > minDistance) {
                        mPrevX.put(id, x);
                        mPrevY.put(id, y);
                    } else {
                        continue;
                    }

                    View newView = findViewByPointer(ev, i);
                    if (newView != null) {
                        for (Iterator<View> it = mTouchedViews.keySet().iterator(); it.hasNext(); ) {
                            View oldView = it.next();
                            if (oldView != newView) {
                                if (dispatchTouchUp(oldView, ev)) {
                                    it.remove();
                                }
                            }
                        }
                        dispatchTouchDown(newView, ev);
                    }
                }
                break;
        }
        return true;
    }

    private void dispatchTouchDown(View v, MotionEvent ev) {
        if (v != null) {
            if (mTouchedViews.get(v) == null) {
                MotionEvent e = MotionEvent.obtain(ev.getDownTime(), ev.getEventTime(),
                        MotionEvent.ACTION_DOWN,
                        ev.getX(ev.getActionIndex()), ev.getY(ev.getActionIndex()), 0);
                mTouchedViews.put(v, new HashSet<>());
                v.dispatchTouchEvent(e);
            }
            mTouchedViews.get(v).add(ev.getPointerId(ev.getActionIndex()));
        }
    }

    private boolean dispatchTouchUp(View v, MotionEvent ev) {
        int id = ev.getPointerId(ev.getActionIndex());
        if (v != null && mTouchedViews.get(v) != null && mTouchedViews.get(v).contains(id)) {
            mTouchedViews.get(v).remove(id);
            if (mTouchedViews.get(v).isEmpty()) {
                MotionEvent e = MotionEvent.obtain(ev.getDownTime(), ev.getEventTime(),
                        MotionEvent.ACTION_UP,
                        ev.getX(ev.getActionIndex()), ev.getY(ev.getActionIndex()), 0);
                v.dispatchTouchEvent(e);
                return true;
            }
        }
        return false;
    }

    private View findViewByPointer(MotionEvent ev, int index) {
        int x = (int) (ev.getX(index) + getX());
        int y = (int) (ev.getY(index) + getY());
        return findViewByPos(x, y);
    }

    private View findViewByPos(int x, int y) {
        for (int i = getChildCount() - 1; i >= 0; i--) {
            View v = getChildAt(i);
            int[] pos = new int[2];
            v.getLocationOnScreen(pos);
            Rect rect = new Rect(pos[0], pos[1], pos[0] + v.getWidth(), pos[1] + v.getHeight());
            if (rect.contains(x, y)) {
                return v;
            }
        }
        return null;
    }

    @Override
    public void view() {
        keys(false);
        keys(true);
    }

    private void keys(boolean black) {
        int numWhiteKeys = 0;
        for (int i = 0; i < mKeyCount; i++) {
            if (!BLACK_KEYS[(mFirstKey + i) % 12]) {
                numWhiteKeys++;
            }
        }
        int keyWidth = mWidth / numWhiteKeys;
        int leftOffset = 0;

        for (int i = 0; i < mKeyCount; i++) {
            final int index = (mFirstKey + i) % 12;
            final int offset = leftOffset;
            final int key = mFirstKey + i;
            if (BLACK_KEYS[index] && black) {
                v(MidiKeyView.class, () -> {
                    size(keyWidth-2*dip(3), mHeight/2);
                    x(offset - keyWidth / 2);
                    tag(key);
                    margin(dip(3));
                    backgroundColor(0xff333333);
                });
            }
            if (!BLACK_KEYS[index] && !black) {
                v(MidiKeyView.class, () -> {
                    size(keyWidth - 2 * dip(3), FILL);
                    tag(key);
                    x(offset);
                    margin(dip(3));
                    backgroundColor(0xffe1e1e1);
                });
            }
            if (!BLACK_KEYS[index]) {
                leftOffset = leftOffset + keyWidth;
            }
        }
    }
}

File: android/app/src/main/java/com/naivesound/scherzo/PianoKeyboard.kt
package com.naivesound.scherzo

import android.content.Context
import android.graphics.Rect
import android.util.DisplayMetrics
import android.util.SparseIntArray
import android.view.MotionEvent
import android.view.View

import java.util.HashSet

import trikita.anvil.DSL.*

import trikita.anvil.Anvil
import trikita.anvil.RenderableView

class PianoKeyboard(c: Context) : RenderableView(c) {

    private var mWidth: Int = 0
    private var mHeight: Int = 0

    private val mTouchedViews: MutableMap<View, MutableSet<Int>> = HashMap()
    private val mPrevX = SparseIntArray()
    private val mPrevY = SparseIntArray()

    private val mKeyCount = 12
    private val mFirstKey = 48

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        mWidth = View.MeasureSpec.getSize(widthMeasureSpec)
        mHeight = View.MeasureSpec.getSize(heightMeasureSpec)
        this.setMeasuredDimension(mWidth, mHeight)
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        Anvil.render()
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        return true
    }

    override fun onTouchEvent(ev: MotionEvent): Boolean {
        val action = ev.actionMasked
        when (action) {
            MotionEvent.ACTION_POINTER_DOWN, MotionEvent.ACTION_DOWN -> {
                val pointerId = ev.getPointerId(ev.actionIndex)
                mPrevX.put(pointerId, ev.getX(ev.actionIndex).toInt())
                mPrevY.put(pointerId, ev.getY(ev.actionIndex).toInt())
                dispatchTouchDown(findViewByPointer(ev, ev.actionIndex), ev)
            }
            MotionEvent.ACTION_OUTSIDE, MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_POINTER_UP, MotionEvent.ACTION_UP -> run {
                val it = mTouchedViews.keys.iterator()
                while (it.hasNext()) {
                    val v = it.next()
                    if (dispatchTouchUp(v, ev)) {
                        it.remove()
                    }
                }
            }
            MotionEvent.ACTION_MOVE -> for (i in 0..ev.pointerCount - 1) {
                val id = ev.getPointerId(i)
                val x = ev.getX(i).toInt()
                val y = ev.getY(i).toInt()
                val minDistance = (24 * (resources.displayMetrics.densityDpi / DisplayMetrics.DENSITY_DEFAULT)).toInt()
                if (Math.abs(mPrevX.get(id, x) - x) > minDistance || Math.abs(mPrevY.get(id, y) - y) > minDistance) {
                    mPrevX.put(id, x)
                    mPrevY.put(id, y)
                } else {
                    continue
                }

                val newView = findViewByPointer(ev, i)
                if (newView != null) {
                    val it = mTouchedViews.keys.iterator()
                    while (it.hasNext()) {
                        val oldView = it.next()
                        if (oldView !== newView) {
                            if (dispatchTouchUp(oldView, ev)) {
                                it.remove()
                            }
                        }
                    }
                    dispatchTouchDown(newView, ev)
                }
            }
        }
        return true
    }

    private fun dispatchTouchDown(v: View?, ev: MotionEvent) {
        if (v != null) {
            if (mTouchedViews[v] == null) {
                val e = MotionEvent.obtain(ev.downTime, ev.eventTime,
                        MotionEvent.ACTION_DOWN,
                        ev.getX(ev.actionIndex), ev.getY(ev.actionIndex), 0)
                mTouchedViews.put(v, HashSet<Int>())
                v.dispatchTouchEvent(e)
            }
            mTouchedViews[v]!!.add(ev.getPointerId(ev.actionIndex))
        }
    }

    private fun dispatchTouchUp(v: View?, ev: MotionEvent): Boolean {
        val id = ev.getPointerId(ev.actionIndex)
        if (v != null && mTouchedViews[v] != null && mTouchedViews[v]!!.contains(id)) {
            mTouchedViews[v]!!.remove(id)
            if (mTouchedViews[v]!!.isEmpty()) {
                val e = MotionEvent.obtain(ev.downTime, ev.eventTime,
                        MotionEvent.ACTION_UP,
                        ev.getX(ev.actionIndex), ev.getY(ev.actionIndex), 0)
                v.dispatchTouchEvent(e)
                return true
            }
        }
        return false
    }

    private fun findViewByPointer(ev: MotionEvent, index: Int): View? {
        val x = (ev.getX(index) + x).toInt()
        val y = (ev.getY(index) + y).toInt()
        return findViewByPos(x, y)
    }

    private fun findViewByPos(x: Int, y: Int): View? {
        for (i in childCount - 1 downTo 0) {
            val v = getChildAt(i)
            val pos = IntArray(2)
            v.getLocationOnScreen(pos)
            val rect = Rect(pos[0], pos[1], pos[0] + v.width, pos[1] + v.height)
            if (rect.contains(x, y)) {
                return v
            }
        }
        return null
    }

    override fun view() {
        keys(false)
        keys(true)
    }

    private fun keys(black: Boolean) {
        var numWhiteKeys = 0
        for (i in 0..mKeyCount - 1) {
            if (!BLACK_KEYS[(mFirstKey + i) % 12]) {
                numWhiteKeys++
            }
        }
        val keyWidth = mWidth / numWhiteKeys
        var leftOffset = 0

        for (i in 0..mKeyCount - 1) {
            val index = (mFirstKey + i) % 12
            val offset = leftOffset
            val key = mFirstKey + i
            if (BLACK_KEYS[index] && black) {
                v(MidiKeyView::class.java, {
                    size(keyWidth - 2 * dip(3), mHeight / 2)
                    x((offset - keyWidth / 2).toFloat())
                    tag(key)
                    margin(dip(3))
                    backgroundColor(0xff333333.toInt())
                })
            }
            if (!BLACK_KEYS[index] && !black) {
                v(MidiKeyView::class.java, {
                    size(keyWidth - 2 * dip(3), FILL)
                    tag(key)
                    x(offset.toFloat())
                    margin(dip(3))
                    backgroundColor(0xffe1e1e1.toInt())
                })
            }
            if (!BLACK_KEYS[index]) {
                leftOffset = leftOffset + keyWidth
            }
        }
    }

    companion object {

        private val TAG = "PianoKeyboard"

        internal var BLACK_KEYS = booleanArrayOf(false, true, false, true, false, false, true, false, true, false, true, false)
    }
}
